frontend/purposeful-frontend/components/Registration.js

```diff
@@ -14,11 +14,93 @@ import {
 } from "@chakra-ui/react";
 import { ViewIcon, ViewOffIcon } from "@chakra-ui/icons";
 import { Field, Form, Formik } from "formik";
+import fetchWrapper, { getAuthorities } from "../utils/fetch_wrapper";
+import notification from "../utils/notification";
+
+async function handleRegistrationForm(values, actions, GrantedAuth) {
+  const payload = {
+    email: values.email,
+    password: values.password,
+    firstname: values.firstname,
+    lastname: values.lastname,
+  };
+  let response = null;
+
+  if (
+    GrantedAuth.length === 0 &&
+    !window.location.pathname.includes("moderator")
+  ) {
+    // User is not authenticated and is registering as a regular user
+    response = await fetchWrapper(
+      "/api/appuser/regular",
+      null,
+      "POST",
+      payload
+    );
+  } else if (
+    GrantedAuth.includes("Owner") &&
+    window.location.pathname.includes("moderator")
+  ) {
+    // User is an admin and is registering a moderator
+    response = await fetchWrapper(
+      "/api/appuser/moderator",
+      null,
+      "POST",
+      payload
+    );
+  } else {
+    // User does not have the required permissions
+    notification(
+      "error",
+      "You do not have the required permissions to perform this action.",
+      null
+    );
+    actions.setSubmitting(false);
+  }
+
+  if (
+    response.ok &&
+    GrantedAuth.includes("Owner") &&
+    window.location.pathname.includes("moderator")
+  ) {
+    // User is an admin and successfully registered a moderator
+    notification("success", "Moderator account created successfully.", null);
+    actions.setSubmitting(false);
+  } else if (response.ok) {
+    // User successfully registered as a regular user
+    actions.setSubmitting(false);
+    window.location.href = "/login"; // Redirect to login page
+  } else if (response !== null) {
+    // User registration failed display error mesages
+    notification("error", "An error occurred.", response.errorMessages);
+    actions.setErrors({
+      firstname: response.errorMessages.toLowerCase().includes("first name")
+        ? response.errorMessages
+        : null,
+      lastname: response.errorMessages.toLowerCase().includes("last name")
+        ? response.errorMessages
+        : null,
+      email: response.errorMessages.toLowerCase().includes("email")
+        ? response.errorMessages
+        : null,
+      password: response.errorMessages.toLowerCase().includes("password")
+        ? response.errorMessages
+        : null,
+    });
+    actions.setSubmitting(false);
+  }
+}
 
 export default function Registration() {
   const [showPassword, setShowPassword] = useState(false);
   const [showConfirmPassword, setShowConfirmPassword] = useState(false);
 
+  // User and authentication info
+  const [GrantedAuth, setGrantedAuth] = useState([]);
```

Comment:
What does this do? Sorry :)

Reply:
This is the useState and useEffect from React. I had to use them because access to the localStorage is only client side, so it would have been `null`